### ISSUE TITLE
Lint Suggestion: Pointer offset with vec reference

### ISSUE THREAD

**Issue by BigWingBeat** [rust-lang/rust-clippy#14943]
### What it does

Fires upon calling one of the pointer `offset_from` methods (including `byte_...` and `..._unsigned`) when the `origin` parameter is `&Vec<T>` or `&mut Vec<T>`.

### Advantage

Prevents UB.
The user likely intended to call `.as_ptr()` on the vec before passing it in, but forgetting to do so results in the reference (to the vec variable, not the vec's own pointer) being silently converted to a pointer instead, which will of course have a completely different address and provenance than the heap allocation they likely intended to use.

### Drawbacks

May or may not have false positives when working with nested vecs, where the given vec reference is itself part of a heap allocation, rather than being a local variable.

### Example

```rust
fn mutate_vec(vec: &mut Vec<i32>) {
    vec.push(9);
    let nine = vec.last().unwrap();
    let ptr = std::ptr::from_ref(nine);
    let offset = unsafe { ptr.byte_offset_from_unsigned(vec) }; // !! UB !!
}
```

Could be written as:

```rust
fn mutate_vec(vec: &mut Vec<i32>) {
    vec.push(9);
    let nine = vec.last().unwrap();
    let ptr = std::ptr::from_ref(nine);
    let offset = unsafe { ptr.byte_offset_from_unsigned(vec.as_ptr()) }; // All good
}
```

**Comment by lukaslueg** [comment]
In principle the same applies to other fat containers like `Arc`, `Rc`
